### ISSUE TITLE
'from' replaced with 'integrity' in npm-shrinkwrap.json, breaking git packages

### ISSUE THREAD

**Issue by Torgen** [meteor/meteor#9477]
Version: 1.6.0.0
Worked in: 1.4.2.3
OS: Ubuntu 16.04LTS
Reproduction:
git clone --branch bad-shrinkwrap https://github.com/Torgen/codex-blackboard bad-blackboard
cd bad-blackboard
meteor build ../bad
\# Will fail on a 404 fetching the codex-blackboard-hubot-scripts package
cd ..
git clone --branch good-shrinkwrap https://github.com/Torgen/codex-blackboard good-blackboard
cd good-blackboard
meteor build ../good
\# Will correctly fetch codex-blackboard-hubot-scripts from github. 
git diff
\# 'from' keys in npm-shrinkwrap.json have been replaced with 'integrity'
diff packages/codex-blackboard-hubot-scripts/.npm/package/npm-shrinkwrap.json ../bad-blackboard/packages/codex-blackboard-hubot-scripts/.npm/package/npm-shrinkwrap.json
\# Files will be more or less the same--the npm-shrinkwrap.json in bad-shrinkwrap was modified from the version in good-shrinkwrap by meteor!

Expected behavior: Meteor doesn't modify my npm-shrinkwrap.json file in a way that I can't commit to my repository.

Note that if you ask meteor to build from bad-blackboard after building in good-blackboard, it will work--as a result, you may wish to test this in a VM that you can reset to a pristine state.

It appears that if 'from' is present, it will be used to determine where to fetch the package from, but 'resolved' will not, meaning npm will try to download a git package from the npm registry.

**Comment by klaussner** [comment]
This looks like a bug in Meteor's shrinkwrap function!

The `"from"` property for the `codex-blackboard-hubot-scripts` package in your "good" `npm-shrinkwrap.json` file has this value:

```
"codex-blackboard-hubot-scripts@https://github.com/cscott/codex-blackboard-hubot-scripts/tarball/62d35c1a0134fc59916cc4c8521f4837cb3135a0"
```

When Meteor calls `npm install`, this value is included in the installed `package.json` file (as `"_from"` in `node_modules/codex-blackboard-hubot-scripts/package.json`). After the dependencies are installed, the `npm-shrinkwrap.json` file is generated from all `package.json` files in `node_modules` (with [`getInstalledDependenciesTree`](https://github.com/meteor/meteor/blob/3f8c90a29f26ec6dc8ff55f736c36a68fa40398f/tools/isobuild/meteor-npm.js#L904)). However, Meteor doesn't handle the `"_from"` property correctly because it doesn't recognize it as a URL (due to the `codex-blackboard-hubot-scripts@` prefix). Therefore, it doesn't use the URL for the `"version"` property. This is the generated shrinkwrap:

```
{
  ...
  "codex-blackboard-hubot-scripts": {
    "version": "2.0.0",
    "resolved": "https://github.com/cscott/codex-blackboard-hubot-scripts/tarball/62d35c1a0134fc59916cc4c8521f4837cb3135a0",
    "integrity": "sha1-ltE65ni59HBf51YXwcVfRV0on0o=",
    ...
```

But it should look like this:

```
...
{
  ...
  "codex-blackboard-hubot-scripts": {
    "version": "https://github.com/cscott/codex-blackboard-hubot-scripts/tarball/62d35c1a0134fc59916cc4c8521f4837cb3135a0",
    "integrity": "sha1-ltE65ni59HBf51YXwcVfRV0on0o=",
    ...
```

**Workaround:** Remove the `codex-blackboard-hubot-scripts@` prefix from the `"from"` field in your `npm-shrinkwrap.json` file and Meteor should update the shrinkwrap correctly.

**Comment by benjamn** [comment]
This should be fixed in `1.6.1-rc.4`, by my commit c84561dfdaf9b47771cd872790738dc1965113d8. Thanks for the report and reproduction @Torgen, and thanks for the accurate diagnosis @klaussner!
